lib/workspace.js

```diff
@@ -336,17 +337,53 @@ module.exports = {
     // local workspace files
     const questionBasePath = path.join(course_path, 'questions', question.qid);
     const localPath = path.join(questionBasePath, 'workspace');
+    const templatePath = path.join(questionBasePath, 'workspaceTemplates');
 
     // base workspace directory wherever we are uploading to
     const remoteDirName = `workspace-${workspace_id}-${workspace.version}`;
     const remotePath = path.join(remoteDirName, 'current');
 
-    // Check if we have any local files to upload
-    const localPathExists = await fse.pathExists(localPath);
+    const staticFiles = (
+      await async
+        .mapSeries(klaw(localPath), async (file) => {
+          return file.stats.isFile()
+            ? { name: path.relative(localPath, file.path), localPath: file.path }
+            : null;
+        })
+        .catch(() => {
+          // Template path does not exist or is not accessible, do nothing
+          return [];
+        })
+    ).filter((file) => file);
+
+    const mustacheParams = { params: variant.params, correct_answers: variant.true_answer };
+
+    const templateFiles = (
+      await async
+        .mapSeries(klaw(templatePath), async (file) => {
+          try {
+            if (!file.stats.isFile()) return null;
+            return {
+              name: path.relative(templatePath, file.path),
+              buffer: mustache.render(
```

Comment:
Per the [Mustache docs](https://github.com/janl/mustache.js/#pre-parsing-and-caching-templates), Mustache maintains an internal cache of parsed templates. For long-running servers, this is not ideal, as this is essentially a memory leak. I think we should disable that by setting `mustache.templateCache = undefined` as described here: https://github.com/janl/mustache.js/pull/731/files. We can probably do this in `server.js` as part of the initialization code.

We probably should have been doing this the whole time we've used mustache - oops!

Related to mustache rendering, I've recently been hyper-aware of the ability of poorly-written or malicious user input to trigger denial-of-service attacks. In the future, we may consider moving potentially-expensive synchronous operations like this to worker threads, or to write a custom renderer based on `mustache.Writer` that can enforce a max size and abort rendering if we reach it. That's all for the future though.

Reply:
Moved to #7098 .